libcontainer/cgroups/fs/memory.go

```diff
@@ -55,9 +58,50 @@ func (s *MemoryGroup) Apply(d *cgroupData) (err error) {
 }
 
 func (s *MemoryGroup) SetKernelMemory(path string, cgroup *configs.Cgroup) error {
-	// This has to be done separately because it has special constraints (it
-	// can't be done after there are processes attached to the cgroup).
-	if cgroup.Resources.KernelMemory > 0 {
+	// This has to be done separately because it has special
+	// constraints (it can only be initialized before setting up a
+	// hierarchy or adding a task to the cgroups. However, if
+	// sucessfully initialized, it can be updated anytime afterwards)
+	if cgroup.Resources.KernelMemory != 0 {
+		kmemInitialized := false
+		// Is kmem.limit_in_bytes already set?
+		if b, err := ioutil.ReadFile(filepath.Join(path, "memory.kmem.limit_in_bytes")); err != nil {
+			return err
+		} else {
+			kmemValue, err := strconv.ParseUint(strings.TrimSpace(string(b)), 10, 64)
+			if err != nil {
+				return err
+			}
+			switch system.GetLongBit() {
+			case 32:
+				kmemInitialized = uint32(kmemValue) != uint32(math.MaxUint32)
+			case 64:
+				kmemInitialized = kmemValue != uint64(math.MaxUint64)
+			}
+		}
+
+		if !kmemInitialized {
```

Comment:
@mlaventure The reason I added this restriction is because I found it's restriction from kernel https://github.com/torvalds/linux/blob/master/Documentation/cgroup-v1/memory.txt#L284-L286

But after more tests and code review, turns out it's a false alarm, the restriction is no longer in kernel for a long time, I already sent a patch to fix this doc in kernel. We can remove this restriction on Docker side once this PR merged, thanks.